src/Aspire.Hosting/Dcp/KubernetesService.cs

```diff
@@ -280,16 +285,78 @@ public void Dispose()
 
     private static bool IsRetryable(Exception ex) => ex is HttpRequestException || ex is KubeConfigException;
 
-    private void EnsureKubernetes()
+    private readonly SemaphoreSlim _kubeconfigReadSemaphore = new(1);
+
+    private async Task EnsureKubernetesAsync(CancellationToken cancellationToken = default)
     {
-        if (_kubernetes != null) { return; }
+        // Return early before waiting for the semaphore if we can.
+        if (_kubernetes != null)
+        {
+            logger.LogDebug("Kubernetes ensured at first chance shortcut.");
+            return;
+        }
+
+        while (!await _kubeconfigReadSemaphore.WaitAsync(100, cancellationToken).ConfigureAwait(false))
+        {
+            logger.LogDebug("Waiting for semaphore to read kubeconfig.");
+        }
```

Comment:
Why the 100ms timeout?

Reply:
Just so we don't have a bunch of threads spinning in a tight loop.

I wanted to log that some spinning was occurring to make diagnostics easier hence why I didn't just specify -1

Reply:
Consider an ETW event for this.